balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java

```diff
@@ -95,43 +90,50 @@ public BalanceComputationImpl(Network network, Map<NetworkArea, Double> networkA
         network.getVariantManager().cloneVariant(workingStateId, workingVariantCopyId);
         network.getVariantManager().setWorkingVariant(workingVariantCopyId);
 
-        Map<NetworkArea, Double> networkAreasResidue;
-        List<NetworkArea> unbalancedNetworkAreas;
+        Map<NetworkArea, Double> mismatches = new HashMap<>();
+        double totalMismatches;
 
-        while (iterationCounter < parameters.getMaxNumberIterations()
-                && result.getStatus() == BalanceComputationResult.Status.FAILED) {
+        do {
+            // Step 2: Perform the scaling
+            for (NetworkArea na : mismatches.keySet()) {
+                double asked = mismatches.get(na);
 
-            // Step 2 : compute Loadflow
+                Scalable scalable = networkAreasScalableMap.get(na);
+                double done = scalable.scale(network, mismatches.get(na));
+                LOGGER.debug("Scaling for area {}: asked={}, done={}", na.getName(), asked, done);
+            }
+
+            // Step 3: compute Loadflow
             LoadFlowResult loadFlowResult = loadFlowRunner.run(network, workingVariantCopyId, computationManager, new LoadFlowParameters());
             if (!loadFlowResult.isOk()) {
                 LOGGER.error("Loadflow on network {} does not converge", network.getId());
                 result = new BalanceComputationResult(BalanceComputationResult.Status.FAILED, iterationCounter);
                 return CompletableFuture.completedFuture(result);
             }
 
-            // Step 3 : Balance computation iteration
-            iterationCounter++;
-            networkAreasResidue = getNetworkAreasResidue(network);
-            unbalancedNetworkAreas = listUnbalancedNetworkAreas(parameters, networkAreasResidue);
-
-            if (unbalancedNetworkAreas.isEmpty()) {
-                // Change the workingStateId with final scaling
-                network.getVariantManager().setWorkingVariant(workingStateId);
-                scaleBalancedNetwork(previousScalingMap);
-                loadFlowRunner.run(network, workingStateId, computationManager, new LoadFlowParameters());
-                result = new BalanceComputationResult(BalanceComputationResult.Status.SUCCESS, iterationCounter, previousScalingMap);
+            // Step 4: Compute balance and mismatch for each area
+            totalMismatches = 0.0;
+            for (NetworkArea na : networkAreaNetPositionTargetMap.keySet()) {
+                double target = networkAreaNetPositionTargetMap.get(na);
+                double balance = na.getNetPosition(network);
+                double oldMismatch = mismatches.computeIfAbsent(na, k -> 0.0);
+                double mismatch = target - balance;
+                mismatches.put(na, oldMismatch + mismatch);
+                LOGGER.debug("Mistmatch for area {}: {} (target={}, balance={})", na.getName(), mismatch, target, balance);
+
+                totalMismatches += mismatch * mismatch;
```

Comment:
While reading this part of the alogrithm, I think the map ```mismatches``` should be renamed, cause it does not really help understanding its content.
Particularly, the line ```mismatches.put(na, oldMismatch + mismatch);``` bugged me ...
Perhaps something like ```shiftForecasts``` or ```shiftEstimates``` ?